File: bindings/heeczer-java/src/test/java/io/github/cognizhi/heeczer/ContractTest.java
package io.github.cognizhi.heeczer;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Contract tests for plan 0001 / ADR-0002 — Java binding.
 *
 * <p>
 * Verifies:
 * <ol>
 * <li>Every valid fixture round-trips through the typed {@link Event} POJO
 * without data loss (parse → serialize → re-parse == original).</li>
 * <li>Extension fields under {@code meta.extensions} survive a round-trip.</li>
 * <li>{@code FAIL_ON_UNKNOWN_PROPERTIES = true} rejects unknown top-level
 * fields.</li>
 * </ol>
 */
class ContractTest {

    /**
     * Lenient mapper matching HeeczerClient: NON_NULL, accept unknown (for open
     * parsing).
     */
    private static final ObjectMapper LENIENT = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .setSerializationInclusion(JsonInclude.Include.NON_NULL);

    /**
     * Strict mapper: rejects unknown properties; used for structural enforcement
     * tests.
     */
    private static final ObjectMapper STRICT = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true)
            .setSerializationInclusion(JsonInclude.Include.NON_NULL);

    private static Path fixtureDir() {
        // Maven runs tests from the module root (bindings/heeczer-java).
        // Walk up two directories to reach the repo root, then down to core/schema.
        return Paths.get("../../core/schema/fixtures/events/valid").toAbsolutePath();
    }

    static Stream<Path> validFixtures() throws IOException {
        Path dir = fixtureDir();
        return Files.list(dir)
                .filter(p -> p.toString().endsWith(".json"))
                .sorted();
    }

    // ── Round-trip tests ──────────────────────────────────────────────────────

    @Test
    void atLeastOneValidFixtureExists() throws IOException {
        assertTrue(validFixtures().findAny().isPresent(),
                "No valid fixtures found in " + fixtureDir());
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("validFixtures")
    void validFixtureRoundTripsLosslessly(Path fixture) throws IOException {
        String body = Files.readString(fixture);

        // Parse with lenient mapper so optional fields are tolerated.
        Event event = LENIENT.readValue(body, Event.class);

        // Re-serialize with NON_NULL (matches Rust skip_serializing_if behavior).
        String reserialised = LENIENT.writeValueAsString(event);

        // Semantic comparison as JsonNode tree.
        JsonNode original = LENIENT.readTree(body);
        JsonNode roundtripped = LENIENT.readTree(reserialised);

        assertEquals(original, roundtripped,
                "Round-trip mismatch for " + fixture.getFileName());
    }

    // ── Extensions round-trip ─────────────────────────────────────────────────

    @Test
    void metaExtensionsSurviveRoundTrip() throws JsonProcessingException {
        Event event = new Event();
        event.specVersion = "1.0";
        event.eventId = "00000000-0000-4000-8000-aabbccddeeff";
        event.timestamp = "2026-04-22T10:00:00Z";
        event.frameworkSource = "test";
        event.workspaceId = "ws_ext";
        event.task = new EventTask("ext_test", Outcome.success);
        event.metrics = new EventMetrics(100);

        EventMeta meta = new EventMeta("java", "0.1.0");
        meta.extensions = LENIENT.readTree("{\"custom_key\":42,\"nested\":{\"x\":true}}");
        event.meta = meta;

        String serialised = LENIENT.writeValueAsString(event);
        Event back = LENIENT.readValue(serialised, Event.class);

        assertNotNull(back.meta.extensions, "meta.extensions must survive round-trip");
        assertEquals(42, back.meta.extensions.get("custom_key").asInt());
        assertTrue(back.meta.extensions.get("nested").get("x").asBoolean());
    }

    @Test
    void absentOptionalFieldsRemainAbsentAfterRoundTrip() throws JsonProcessingException {
        Event event = new Event();
        event.specVersion = "1.0";
        event.eventId = "00000000-0000-4000-8000-000000000001";
        event.timestamp = "2026-04-22T10:00:00Z";
        event.frameworkSource = "test";
        event.workspaceId = "ws_min";
        event.task = new EventTask("min_task", Outcome.success);
        event.metrics = new EventMetrics(50);
        event.meta = new EventMeta("java", "0.1.0");

        String serialised = LENIENT.writeValueAsString(event);
        JsonNode node = LENIENT.readTree(serialised);

        assertFalse(node.has("correlation_id"), "correlation_id must be absent");
        assertFalse(node.has("identity"), "identity must be absent");
        assertFalse(node.has("context"), "context must be absent");
        assertFalse(node.get("meta").has("extensions"), "meta.extensions must be absent");
    }

    // ── Strict unknown-field rejection ────────────────────────────────────────

    @Test
    void unknownTopLevelFieldRejectedByStrictMapper() {
        String bad = """
                {
                  "spec_version": "1.0",
                  "event_id": "00000000-0000-4000-8000-aabbccddeeff",
                  "timestamp": "2026-04-22T10:00:00Z",
                  "framework_source": "test",
                  "workspace_id": "ws_strict",
                  "task": {"name": "t", "outcome": "success"},
                  "metrics": {"duration_ms": 100},
                  "meta": {"sdk_language": "java", "sdk_version": "0.1.0"},
                  "forbidden_extra_field": "value"
                }
                """;

        assertThrows(UnrecognizedPropertyException.class,
                () -> STRICT.readValue(bad, Event.class),
                "Strict mapper must reject unknown top-level field");
    }

    @Test
    void unknownMetaFieldRejectedByStrictMapper() {
        String bad = """
                {
                  "spec_version": "1.0",
                  "event_id": "00000000-0000-4000-8000-aabbccddeeff",
                  "timestamp": "2026-04-22T10:00:00Z",
                  "framework_source": "test",
                  "workspace_id": "ws_strict",
                  "task": {"name": "t", "outcome": "success"},
                  "metrics": {"duration_ms": 100},
                  "meta": {
                    "sdk_language": "java",
                    "sdk_version": "0.1.0",
                    "unknown_meta_key": "oops"
                  }
                }
                """;

        assertThrows(UnrecognizedPropertyException.class,
                () -> STRICT.readValue(bad, Event.class),
                "Strict mapper must reject unknown field inside meta (use meta.extensions instead)");
    }
}

File: bindings/heeczer-js/tests/contract.test.ts
/**
 * Contract tests for plan 0001 / ADR-0002 — TypeScript/JavaScript binding.
 *
 * Verifies:
 * 1. Every valid fixture round-trips through the typed Event interface
 *    without data loss (parse → cast → stringify → re-parse == original).
 * 2. Extension fields under meta.extensions survive a round-trip.
 *
 * Note: Unknown top-level field enforcement is provided by the server-side
 * JSON Schema validator (Event interface is structural in TypeScript).
 * The TypeScript type system rejects extra properties in object literals
 * at compile time — see the @ts-expect-error block below.
 */

import { describe, expect, it } from "vitest";
import {
  HeeczerValidationError,
  validateEvent,
  type Event,
} from "../src/index.js";

// Load all valid fixture files as raw strings via Vite's glob import.
// Keys are relative paths; values are raw JSON strings.
// import.meta.glob is a Vite-specific runtime feature; cast via unknown to
// avoid requiring vite/client type definitions in devDependencies.
type GlobFn = (
  pattern: string,
  opts: { query: string; import: string; eager: boolean },
) => Record<string, string>;
const fixtureRaw = (import.meta as unknown as { glob: GlobFn }).glob(
  "../../../core/schema/fixtures/events/valid/*.json",
  {
    query: "?raw",
    import: "default",
    eager: true,
  },
);

function loadValidFixtures(): Array<{ name: string; body: string }> {
  return Object.entries(fixtureRaw)
    .map(([path, body]) => ({
      name: path.split("/").at(-1) ?? path,
      body,
    }))
    .sort((a, b) => a.name.localeCompare(b.name));
}

describe("Event contract: valid fixture round-trips", () => {
  const fixtures = loadValidFixtures();

  it("loads at least one valid fixture", () => {
    expect(fixtures.length).toBeGreaterThan(0);
  });

  for (const { name, body } of fixtures) {
    it(`round-trips ${name} losslessly`, () => {
      // Parse the fixture as an Event (TypeScript cast — no runtime stripping).
      const event = JSON.parse(body) as Event;
      validateEvent(event);

      // Re-serialize and re-parse for semantic comparison.
      const roundTripped = JSON.parse(JSON.stringify(event));
      const original = JSON.parse(body);

      // Must be semantically equal (key order may differ).
      expect(roundTripped).toEqual(original);
    });
  }
});

describe("Event contract: extensions round-trip", () => {
  it("meta.extensions survives JSON round-trip", () => {
    const event: Event = {
      spec_version: "1.0",
      event_id: "00000000-0000-4000-8000-aabbccddeeff",
      timestamp: "2026-04-22T10:00:00Z",
      framework_source: "test",
      workspace_id: "ws_ext",
      task: { name: "ext_test", outcome: "success" },
      metrics: { duration_ms: 100 },
      meta: {
        sdk_language: "node",
        sdk_version: "0.1.0",
        extensions: { custom_key: 42, nested: { x: true } },
      },
    };

    const roundTripped = JSON.parse(JSON.stringify(event)) as Event;

    expect(roundTripped.meta.extensions).toBeDefined();
    expect(
      (roundTripped.meta.extensions as Record<string, unknown>)["custom_key"],
    ).toBe(42);
    expect(
      (
        (roundTripped.meta.extensions as Record<string, unknown>)[
          "nested"
        ] as Record<string, unknown>
      )["x"],
    ).toBe(true);
  });

  it("absent optional fields remain absent after round-trip", () => {
    const event: Event = {
      spec_version: "1.0",
      event_id: "00000000-0000-4000-8000-000000000001",
      timestamp: "2026-04-22T10:00:00Z",
      framework_source: "test",
      workspace_id: "ws_min",
      task: { name: "min_task", outcome: "success" },
      metrics: { duration_ms: 50 },
      meta: { sdk_language: "node", sdk_version: "0.1.0" },
    };

    const roundTripped = JSON.parse(JSON.stringify(event)) as Event;
    // Optional fields not set must not appear.
    expect(roundTripped.correlation_id).toBeUndefined();
    expect(roundTripped.identity).toBeUndefined();
    expect(roundTripped.context).toBeUndefined();
    expect(roundTripped.meta.extensions).toBeUndefined();
  });
});

describe("Event contract: TypeScript unknown-field rejection", () => {
  it("TypeScript type system rejects unknown properties at compile time", () => {
    // The @ts-expect-error below is validated by `pnpm typecheck` (tsc --noEmit).
    // If TypeScript catches the excess property, the directive suppresses the
    // error cleanly. If TypeScript stops catching it, tsc will error on the
    // directive itself — making the compile-time guarantee self-enforcing.
    const _bad: Event = {
      spec_version: "1.0",
      event_id: "00000000-0000-4000-8000-000000000002",
      timestamp: "2026-04-22T10:00:00Z",
      framework_source: "test",
      workspace_id: "ws_ts",
      task: { name: "t", outcome: "success" },
      metrics: { duration_ms: 1 },
      meta: { sdk_language: "node", sdk_version: "0.1.0" },
      // @ts-expect-error – forbidden_extra_field is not in Event (excess property check)
      forbidden_extra_field: "value",
    };
    void _bad; // prevent unused variable warning
    expect(true).toBe(true);
  });

  it("runtime validator rejects unknown top-level fields before transport", () => {
    expect(() =>
      validateEvent({
        spec_version: "1.0",
        event_id: "00000000-0000-4000-8000-000000000002",
        timestamp: "2026-04-22T10:00:00Z",
        framework_source: "test",
        workspace_id: "ws_ts",
        task: { name: "t", outcome: "success" },
        metrics: { duration_ms: 1 },
        meta: { sdk_language: "node", sdk_version: "0.5.1" },
        forbidden_extra_field: "value",
      }),
    ).toThrow(HeeczerValidationError);
  });
});
